pkg/controller/restore_controller.go

```diff
@@ -222,25 +257,45 @@ func (r *restoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 
 func (r *restoreReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		WithEventFilter(kubeutil.NewCreateEventPredicate(func(obj client.Object) bool {
-			restore := obj.(*api.Restore)
-
-			switch restore.Status.Phase {
-			case "", api.RestorePhaseNew:
+		For(&api.Restore{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(ce event.CreateEvent) bool {
 				// only process new restores
-				return true
-			default:
-				r.logger.WithFields(logrus.Fields{
-					"restore": kubeutil.NamespaceAndName(restore),
-					"phase":   restore.Status.Phase,
-				}).Debug("Restore is not new, skipping")
+				return createPredicate(ce)
+			},
+			UpdateFunc: func(ue event.UpdateEvent) bool {
+				// only process the deletion for restores attached with a finalizer
+				// note that a finalizer makes a deletion on the object become an update by setting deletion timestamp
+				return updatePredicate(ue)
+			},
+			DeleteFunc: func(de event.DeleteEvent) bool {
 				return false
-			}
+			},
+			GenericFunc: func(ge event.GenericEvent) bool {
+				return false
+			},
 		})).
-		For(&api.Restore{}).
 		Complete(r)
 }
 
+func createPredicate(ce event.CreateEvent) bool {
+	restore := ce.Object.(*api.Restore)
+	switch restore.Status.Phase {
+	case "", api.RestorePhaseNew:
+		return true
+	default:
+		return false
+	}
+}
+
+func updatePredicate(ue event.UpdateEvent) bool {
+	restore := ue.ObjectNew.(*api.Restore)
+	if !restore.DeletionTimestamp.IsZero() && len(restore.Finalizers) != 0 && (restore.Status.Phase == api.RestorePhaseFailed ||
+		restore.Status.Phase == api.RestorePhasePartiallyFailed || restore.Status.Phase == api.RestorePhaseCompleted) {
```

Comment:
Before this change, a restore can be deleted during the processing, but now it cannot?

Reply:
Yes, it can be deleted during the processing as well after this change. The finalizer is added in terminal phase, so restores in the progress won't get attached with finalizer and they can be deleted directly.